File: v2/16.py





class Solution:
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        # 5 star, 与3sum类似，只是每次不是与0做比较而是与目标数字做比较
        nums.sort()
        rs = nums[0] + nums[1] + nums[2]
        for i in range(len(nums) - 2):
            if i == 0 or nums[i] > nums[i-1]:
                left = i + 1
                right = len(nums) - 1
                while left < right:
                    total = nums[i] + nums[left] + nums[right]
                    if abs(target - total) < abs(target - rs):
                        rs = total
                    if total < target:
                        left += 1
                    elif total > target:
                        right -= 1
                    else:
                        return total
        return rs

File: go/16.go
//package main
//
//import (
//	"math"
//	"sort"
//)

// 也很巧妙，6 star
// 思路同3sum，从左到右遍历，然后对后面的从两边往中间进行逼近，每次比较当前的和目前最接近的和谁更接近
// 然后按照当前和和目标值的相对大小调整左右的索引
func threeSumClosest(nums []int, target int) int {
	sort.Ints(nums)
	length := len(nums)
	gap := nums[0] + nums[1] + nums[2]

	for index, _ := range nums {
		if index == 0 || nums[index] > nums[index-1] {
			left, right := index+1, length-1
			for left < right {
				curSum := nums[index] + nums[right] + nums[left]
				if math.Abs(float64(curSum-target)) < math.Abs(float64(gap-target)) {
					gap = curSum
				}
				if curSum > target {
					right -= 1
				} else if curSum < target {
					left += 1
				} else {
					return curSum
				}
			}
		}

	}
	return gap

}

//func main() {
//	threeSumClosest([]int{0, 0, 0}, 1)
//
//}
